packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts

```diff
@@ -0,0 +1,1156 @@
+  }
+
+  private isAtEnd(): boolean {
+    return (
+      this.current >= this.tokens.length || this.tokens[this.current].kind === SyntaxTokenKind.EOF
+    );
+  }
+
+  private advance(): SyntaxToken {
+    if (this.isAtEnd()) {
+      return _.last(this.tokens)!; // The EOF
+    }
+
+    // eslint-disable-next-line no-plusplus
+    return this.tokens[this.current++];
+  }
+
+  private peek(lookahead: number = 0): SyntaxToken {
+    if (lookahead + this.current >= this.tokens.length) {
+      return _.last(this.tokens)!; // The EOF
+    }
+
+    return this.tokens[this.current + lookahead];
+  }
+
+  private match(...kind: SyntaxTokenKind[]): boolean {
+    const res = this.check(...kind);
+    if (res) {
+      this.advance();
+    }
+
+    return res;
+  }
+
+  private check(...kind: SyntaxTokenKind[]): boolean {
+    const currentToken = this.peek();
+
+    return kind.includes(currentToken.kind);
+  }
+
+  private previous(): SyntaxToken {
+    return this.tokens[this.current - 1];
+  }
+
+  private canHandle<T extends SyntaxNode>(e: PartialParsingError<T>): boolean {
+    return e.handlerContext === undefined || e.handlerContext === this.contextStack.top();
+  }
+
+  private consume(message: string, ...kind: SyntaxTokenKind[]) {
+    if (!this.match(...kind)) {
+      this.logError(this.peek(), CompileErrorCode.UNEXPECTED_TOKEN, message);
+      throw new PartialParsingError(
+        this.peek(),
+        undefined,
+        this.contextStack.findHandlerContext(this.tokens, this.current),
+      );
+    }
+  }
+
+  // Discard tokens until one of `kind` is found
+  // If any tokens are discarded, the error message is logged
+  // Return whether the token of one of the listed kinds are eventually reached
+  private discardUntil(message: string, ...kind: SyntaxTokenKind[]): boolean {
+    if (this.isAtEnd() || !this.check(...kind)) {
+      markInvalid(this.peek());
+      this.logError(this.advance(), CompileErrorCode.UNEXPECTED_TOKEN, message);
+      while (!this.isAtEnd() && !this.check(...kind)) {
+        markInvalid(this.advance());
+      }
+
+      return !this.isAtEnd();
+    }
+
+    return true;
+  }
+
+  private gatherInvalid() {
+    const tokens: SyntaxToken[] = [];
+
+    const firstInvalidList: SyntaxToken[] = [];
+    let curValid: SyntaxToken | undefined;
+    let i = 0;
+    for (; i < this.tokens.length; i += 1) {
+      if (this.tokens[i].isInvalid) {
+        firstInvalidList.push(this.tokens[i]);
+      } else {
+        break;
+      }
+    }
+
+    curValid = this.tokens[i];
+    curValid.leadingInvalid = firstInvalidList;
+    tokens.push(curValid);
+    for (i += 1; i < this.tokens.length; i += 1) {
+      const curToken = this.tokens[i];
+      if (curToken.isInvalid) {
+        curValid.trailingInvalid.push(curToken);
+      } else {
+        curValid = curToken;
+        tokens.push(curValid);
+      }
+    }
+
+    _.remove(this.tokens);
+    this.tokens.push(...tokens);
```

Comment:
Beware of exceeding maximum JS call stack when the number of tokens is too (https://stackoverflow.com/questions/22747068/is-there-a-max-number-of-arguments-javascript-functions-can-accept)